src/main.ts

```diff
@@ -1,19 +1,37 @@
 import { setFailed, setOutput } from '@actions/core'
-import { mkdirP, mv } from '@actions/io/'
+import { mkdirP, mv, cp } from '@actions/io/'
 import { exists } from '@actions/io/lib/io-util'
+import { spawn } from 'child_process'
+import { promisify } from 'util'
 
 import { getVars } from './lib/getVars'
 import { isErrorLike } from './lib/isErrorLike'
 import log from './lib/log'
 
+const rsync = promisify(spawn)
+
 async function main(): Promise<void> {
   try {
     const { cachePath, targetDir, targetPath, options } = getVars()
-
     if (await exists(cachePath)) {
       await mkdirP(targetDir)
-      await mv(cachePath, targetPath, { force: true })
-      log.info(`Cache found and restored to ${options.path}`)
+
+      if (options.copyStrategy === 'move') {
+        await mv(cachePath, targetPath, { force: true })
+      } else if (options.copyStrategy === 'copy') {
+        await cp(cachePath, targetPath, {
+          copySourceDirectory: false,
+          recursive: true,
+          force: true,
+        })
+      } else if (options.copyStrategy === 'rsync') {
+        await rsync('rsync', ['-avqR', '.', targetPath], { stdio: 'inherit', cwd: cachePath })
+      } else {
```

Comment:
1. Correct it does
2. Also correct and I did not address them in this PR
3. Not sure about what you are referring to here. Option to have additional rsync flags?
4. You're right, I forgot that the target path will be empty. It would def improve times on the post step (we could improve the main step by instead cp'ing), but might not be a very common use case.

I agree with your points in that rsync might not be that necessary at all. Will remove for the scope of this PR.

Reply:
Just to clarify, on point number 3 I was highlighting that the options `-avqR` is arbitrary and poses a problem because might not fit everyone's use-case, so the alternative would have been adding an action option that overrides those flags.

As an aside, this is funny:

```
-a, --archive
       This is equivalent to -rlptgoD. It is a quick way  of  saying  you  want  recursion  and  want  to
       preserve  almost  everything  (with -H being a notable omission).  The only exception to the above
       equivalence is when --files-from is specified, in which case -r is not implied.

       Note that -a does not preserve hardlinks, because finding multiply-linked files is expensive.  You
       must separately specify -H.
```

"This is equivalent to `-rlptgoD`" - holy cow! 😆